### ISSUE TITLE
Error handler: access to log message that caused the problem

### ISSUE THREAD

**Issue by D-r-P-3-p-p-3-r** [gabime/spdlog#2046]
Currently, all that's available in the error handler is a description of what went wrong. E.g.

> argument not found

It is not possible to see what log message caused the problem.

This is very annoying when you have a multi-threaded application with a lot of logging. In that case, you can only tell that something went wrong, but it is hard to impossible to figure out which log message caused the problem.
Also, you lose the complete log message, not only the dynamic content. This might lose valuable information about a code path taken when analyzing logs from a customer. Maybe it's a rarely taken code path which is why the devs never noticed the log message was broken...

Can you please extend the error handler in such a way that the bad log message is available along with the error message? Or if that breaks the API, offer an additional error handler with 2 arguments?

Edit:
What would be a bonus: Also, pass the other arguments to the logging call to the error handler.

**Comment by tt4g** [comment]
The macro `FMT_STRING` provided by the {fmt} library checks the number of arguments at compile time.
Doesn't this help you?

See: https://fmt.dev/8.0.0/#safety

**Comment by D-r-P-3-p-p-3-r** [comment]
> The macro `FMT_STRING` provided by the {fmt} library checks the number of arguments at compile time.
> Doesn't this help you?
> 
> See: https://fmt.dev/8.0.0/#safety

I'll check it out. But it seems a little clumsy and verbose.
Does FMT_STRING catch all errors that the error handler would?

And what's the point of the error handler at all if the best practice is to use libfmt checking features?

**Comment by tt4g** [comment]
`FMT_STRING` only works for log messages and arguments formatted with the {fmt} library.
The error handler reports any error, such as a file cannot be opened. It is not specific to log messages.

**Comment by D-r-P-3-p-p-3-r** [comment]
So, if FMT_STRING is the way to go: Why doesn't spdlog automatically apply the macro to format string itself instead of putting that burden on the user of the library?

**Comment by D-r-P-3-p-p-3-r** [comment]
Another variation of passing more information to the error handler.
Instead of passing this to the error handler

> argument not found

spdlog could pass something like this

> argument not found in 'some invalid message to trigger an error {}{}{}{}'

This wouldn't even break the existing interface.

**Comment by gabime** [comment]
> So, if FMT_STRING is the way to go: Why doesn't spdlog automatically apply the macro to format string itself instead of putting that burden on the user of the library

Starting from version 1.9.1, if compiling with the c++20, format strings are validated at compile time without the need of FMT_STRING (not supported by msvc yet afaik).

**Comment by D-r-P-3-p-p-3-r** [comment]
> > So, if FMT_STRING is the way to go: Why doesn't spdlog automatically apply the macro to format string itself instead of putting that burden on the user of the library
> 
> Starting from version 1.9.1, if compiling with the c++20, format strings are validated at compile time without the need of FMT_STRING (not supported by msvc yet afaik).

That is good news. But if you're developing for an embedded platform and are stuck on an older toolchain for a while (or apparently are using MSVC), it's still a problem.

Would it be such a pain to incorporate the problematic log message in the error message?

**Comment by gabime** [comment]
Yes since this is validated by fmt lib, not by spdlog. If you find a way a PR would be most welcome.

**Comment by D-r-P-3-p-p-3-r** [comment]
It took a short look. I think it should be pretty simple.

Just give SPDLOG_LOGGER_CATCH() a parameter for additional infos:

```
#ifndef SPDLOG_NO_EXCEPTIONS
#    define SPDLOG_LOGGER_CATCH(additionalInfo)                                                                                                          \
        catch (const std::exception &ex)                                                                                                   \
        {                                                                                                                                  \
            err_handler_(fmt::format("{} ({})", ex.what(), additionalInfo));                                                                                                       \
        }                                                                                                                                  \
        catch (...)                                                                                                                        \
        {                                                                                                                                  \
            err_handler_("Rethrowing unknown exception in logger");                                                                        \
            throw;                                                                                                                         \
        }
#else
#    define SPDLOG_LOGGER_CATCH(additionalInfo)
#endif
```
The calls to the macro look this then:
```
SPDLOG_LOGGER_CATCH(msg.payload)
SPDLOG_LOGGER_CATCH("")
SPDLOG_LOGGER_CATCH(fmt)
```
Location and log level could also be easily added to the additional info.

The output looks like this then:

```
[*** LOG ERROR #0001 ***] [2021-08-11 14:51:35] [Foo] {argument not found (Boom {}{})}
```

**Comment by gabime** [comment]
Might work. PR would be welcome.

**Comment by D-r-P-3-p-p-3-r** [comment]
> Might work. PR would be welcome.

Done. See https://github.com/gabime/spdlog/pull/2048

**Comment by gabime** [comment]
fixed in #2048
